File: src/client-windows/Pim.Client.App/ViewModels/CalendarViewModel.cs
using System.Collections.ObjectModel;
using System.IO;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Pim.Client.Core.Models;
using Pim.Client.Core.Services;

namespace Pim.Client.App.ViewModels;

public partial class CalendarViewModel : ObservableObject
{
    private readonly ApiClient _api;

    [ObservableProperty] private DateTime _currentMonth = DateTime.Now;
    [ObservableProperty] private ObservableCollection<EventResponse> _events = new();
    [ObservableProperty] private EventResponse? _selectedEvent;
    [ObservableProperty] private bool _isLoading;
    [ObservableProperty] private bool _isEditorOpen;
    [ObservableProperty] private string _editorTitle = string.Empty;
    [ObservableProperty] private string? _editorDescription;
    [ObservableProperty] private string? _editorLocation;
    [ObservableProperty] private DateTime _editorStart = DateTime.Now;
    [ObservableProperty] private DateTime _editorEnd = DateTime.Now.AddHours(1);
    [ObservableProperty] private string _errorMessage = string.Empty;

    public CalendarViewModel(ApiClient api)
    {
        _api = api;
    }

    public int Year => CurrentMonth.Year;
    public int Month => CurrentMonth.Month;
    public int DaysInMonth => DateTime.DaysInMonth(Year, Month);
    public DayOfWeek FirstDayOfWeek => new DateTime(Year, Month, 1).DayOfWeek;
    public int LeadingBlanks => (int)FirstDayOfWeek;

    public async Task LoadEventsAsync()
    {
        IsLoading = true;
        ErrorMessage = string.Empty;
        try
        {
            var start = new DateTimeOffset(Year, Month, 1, 0, 0, 0, TimeSpan.Zero);
            var end = start.AddMonths(1);
            var result = await _api.GetAsync<ApiResponse<List<EventResponse>>>(
                $"/calendar/events?start={start:yyyy-MM-dd}&end={end:yyyy-MM-dd}");
            Events = new ObservableCollection<EventResponse>(result?.Data ?? new List<EventResponse>());
        }
        catch (Exception ex)
        {
            ErrorMessage = $"加载失败: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    [RelayCommand]
    private void PreviousMonth()
    {
        CurrentMonth = CurrentMonth.AddMonths(-1);
        OnPropertyChanged(nameof(Year));
        OnPropertyChanged(nameof(Month));
        OnPropertyChanged(nameof(DaysInMonth));
        OnPropertyChanged(nameof(FirstDayOfWeek));
        OnPropertyChanged(nameof(LeadingBlanks));
    }

    [RelayCommand]
    private void NextMonth()
    {
        CurrentMonth = CurrentMonth.AddMonths(1);
        OnPropertyChanged(nameof(Year));
        OnPropertyChanged(nameof(Month));
        OnPropertyChanged(nameof(DaysInMonth));
        OnPropertyChanged(nameof(FirstDayOfWeek));
        OnPropertyChanged(nameof(LeadingBlanks));
    }

    [RelayCommand]
    private void OpenCreateEditor(DateTime? date = null)
    {
        SelectedEvent = null;
        EditorTitle = string.Empty;
        EditorDescription = null;
        EditorLocation = null;
        EditorStart = date ?? DateTime.Now;
        EditorEnd = EditorStart.AddHours(1);
        IsEditorOpen = true;
    }

    [RelayCommand]
    private void OpenEditEditor(EventResponse evt)
    {
        SelectedEvent = evt;
        EditorTitle = evt.Title;
        EditorDescription = evt.Description;
        EditorLocation = evt.Location;
        EditorStart = evt.DtStart.DateTime;
        EditorEnd = evt.DtEnd.DateTime;
        IsEditorOpen = true;
    }

    [RelayCommand]
    private void CloseEditor()
    {
        IsEditorOpen = false;
    }

    [RelayCommand]
    private async Task SaveEventAsync()
    {
        if (string.IsNullOrWhiteSpace(EditorTitle)) return;
        IsLoading = true;
        try
        {
            var body = new
            {
                calendarId = "", // will use default
                title = EditorTitle,
                description = EditorDescription,
                location = EditorLocation,
                dtStart = new DateTimeOffset(EditorStart),
                dtEnd = new DateTimeOffset(EditorEnd)
            };

            if (SelectedEvent is not null)
            {
                await _api.PutAsync<object>($"/calendar/events/{SelectedEvent.Id}", body);
            }
            else
            {
                // For now, get first calendar
                var cals = await _api.GetAsync<ApiResponse<List<CalendarResponse>>>("/calendar/calendars");
                var calId = cals?.Data?.FirstOrDefault()?.Id;
                if (calId is null)
                {
                    ErrorMessage = "请先创建日历";
                    return;
                }
                await _api.PostAsync<object>("/calendar/events",
                    new { calendarId = calId, title = EditorTitle, description = EditorDescription, location = EditorLocation, dtStart = new DateTimeOffset(EditorStart), dtEnd = new DateTimeOffset(EditorEnd) });
            }

            IsEditorOpen = false;
            await LoadEventsAsync();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"保存失败: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    [RelayCommand]
    private async Task DeleteEventAsync(EventResponse evt)
    {
        try
        {
            await _api.DeleteAsync($"/calendar/events/{evt.Id}");
            await LoadEventsAsync();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"删除失败: {ex.Message}";
        }
    }

    public async Task SyncOutlookAsync()
    {
        IsLoading = true;
        ErrorMessage = string.Empty;
        try
        {
            await _api.PostAsync<object>("/calendar/outlook/sync", new { });
            await LoadEventsAsync();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"同步失败: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    public async Task ImportIcsAsync(string filePath)
    {
        IsLoading = true;
        ErrorMessage = string.Empty;
        try
        {
            var icsContent = await File.ReadAllTextAsync(filePath);
            await _api.PostStringAsync<object>("/calendar/import-ics", icsContent);
            await LoadEventsAsync();
        }
        catch (Exception ex)
        {
            ErrorMessage = $"导入失败: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    public async Task ExportIcsAsync(string filePath)
    {
        IsLoading = true;
        ErrorMessage = string.Empty;
        try
        {
            var start = new DateTimeOffset(Year, Month, 1, 0, 0, 0, TimeSpan.Zero);
            var end = start.AddMonths(1);
            var result = await _api.GetAsync<ApiResponse<string>>(
                $"/calendar/export-ics?start={start:yyyy-MM-dd}&end={end:yyyy-MM-dd}");
            if (result?.Data is not null)
                await File.WriteAllTextAsync(filePath, result.Data);
        }
        catch (Exception ex)
        {
            ErrorMessage = $"导出失败: {ex.Message}";
        }
        finally
        {
            IsLoading = false;
        }
    }

    public List<EventResponse> GetEventsForDay(int day)
    {
        var date = new DateTime(Year, Month, day);
        return Events.Where(e =>
            e.DtStart.Date <= date && e.DtEnd.Date >= date).ToList();
    }
}

File: src/client-android/app/src/main/java/com/pim/app/ui/calendar/CalendarViewModel.kt
package com.pim.app.ui.calendar

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pim.core.models.*
import com.pim.core.network.ApiService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.YearMonth
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.inject.Inject

data class CalendarUiState(
    val currentMonth: YearMonth = YearMonth.now(),
    val events: List<EventResponse> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val isEditorOpen: Boolean = false,
    val selectedEvent: EventResponse? = null,
    val editorTitle: String = "",
    val editorDescription: String = "",
    val editorLocation: String = "",
    val editorStart: String = "",
    val editorEnd: String = "",
    val isImporting: Boolean = false,
    val icsInput: String = ""
)

@HiltViewModel
class CalendarViewModel @Inject constructor(
    private val api: ApiService
) : ViewModel() {

    private val _state = MutableStateFlow(CalendarUiState())
    val state: StateFlow<CalendarUiState> = _state.asStateFlow()

    private val fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss")

    init { loadEvents() }

    fun loadEvents() {
        _state.value = _state.value.copy(isLoading = true, error = null)
        val month = _state.value.currentMonth
        val start = month.atDay(1).atStartOfDay().toInstant(ZoneOffset.UTC).toString()
        val end = month.plusMonths(1).atDay(1).atStartOfDay().toInstant(ZoneOffset.UTC).toString()
        viewModelScope.launch {
            try {
                val res = api.getEvents(start, end)
                if (res.code == 0) {
                    _state.value = _state.value.copy(events = res.data ?: emptyList(), isLoading = false)
                } else {
                    _state.value = _state.value.copy(isLoading = false, error = res.message)
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = "加载失败: ${e.message}")
            }
        }
    }

    fun previousMonth() {
        _state.value = _state.value.copy(currentMonth = _state.value.currentMonth.minusMonths(1))
        loadEvents()
    }

    fun nextMonth() {
        _state.value = _state.value.copy(currentMonth = _state.value.currentMonth.plusMonths(1))
        loadEvents()
    }

    fun openCreateEditor(date: LocalDate? = null) {
        val d = date ?: LocalDate.now()
        val start = d.atTime(9, 0).format(fmt)
        val end = d.atTime(10, 0).format(fmt)
        _state.value = _state.value.copy(
            isEditorOpen = true, selectedEvent = null,
            editorTitle = "", editorDescription = "", editorLocation = "",
            editorStart = start, editorEnd = end
        )
    }

    fun openEditEditor(event: EventResponse) {
        _state.value = _state.value.copy(
            isEditorOpen = true, selectedEvent = event,
            editorTitle = event.title, editorDescription = event.description ?: "",
            editorLocation = event.location ?: "",
            editorStart = event.dtStart, editorEnd = event.dtEnd
        )
    }

    fun closeEditor() { _state.value = _state.value.copy(isEditorOpen = false) }

    fun updateEditorTitle(v: String) { _state.value = _state.value.copy(editorTitle = v) }
    fun updateEditorDescription(v: String) { _state.value = _state.value.copy(editorDescription = v) }
    fun updateEditorLocation(v: String) { _state.value = _state.value.copy(editorLocation = v) }
    fun updateEditorStart(v: String) { _state.value = _state.value.copy(editorStart = v) }
    fun updateEditorEnd(v: String) { _state.value = _state.value.copy(editorEnd = v) }

    fun saveEvent() {
        val s = _state.value
        if (s.editorTitle.isBlank()) return
        _state.value = _state.value.copy(isLoading = true, error = null)
        val body = CreateEventRequest(
            calendarId = "",
            title = s.editorTitle,
            description = s.editorDescription.ifBlank { null },
            location = s.editorLocation.ifBlank { null },
            dtStart = s.editorStart,
            dtEnd = s.editorEnd
        )
        viewModelScope.launch {
            try {
                if (s.selectedEvent != null) {
                    api.updateEvent(s.selectedEvent.id, body)
                } else {
                    val cals = api.getCalendars()
                    val calId = cals.data?.firstOrNull()?.id ?: run {
                        _state.value = _state.value.copy(isLoading = false, error = "请先创建日历")
                        return@launch
                    }
                    api.createEvent(body.copy(calendarId = calId))
                }
                _state.value = _state.value.copy(isLoading = false, isEditorOpen = false)
                loadEvents()
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = "保存失败: ${e.message}")
            }
        }
    }

    fun deleteEvent(id: String) {
        viewModelScope.launch {
            try {
                api.deleteEvent(id)
                loadEvents()
            } catch (e: Exception) {
                _state.value = _state.value.copy(error = "删除失败: ${e.message}")
            }
        }
    }

    fun getEventsForDay(day: Int): List<EventResponse> {
        val month = _state.value.currentMonth
        val date = month.atDay(day)
        return _state.value.events.filter {
            try {
                val start = LocalDate.parse(it.dtStart.substring(0, 10))
                val end = LocalDate.parse(it.dtEnd.substring(0, 10))
                !date.isBefore(start) && !date.isAfter(end)
            } catch (_: Exception) { false }
        }
    }
}
